gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/plan/domain_service/PlanValidatorDomainService.java

```diff
@@ -0,0 +1,110 @@
+        this.policyValidationDomainService = policyValidationDomainService;
+        this.pageCrudService = pageCrudService;
+    }
+
+    public void validatePlanSecurity(Plan plan, String currentOrganizationId, String currentEnvironmentId) {
+        if (plan.getMode().equals(PlanMode.STANDARD)) {
+            ensurePlanSecurityIsAllowed(plan.getSecurity().getType(), currentOrganizationId, currentEnvironmentId);
+            policyValidationDomainService.validateAndSanitizeConfiguration(
+                plan.getSecurity().getType(),
+                plan.getSecurity().getConfiguration()
+            );
+        }
+
+        if (plan.getMode().equals(PlanMode.PUSH) && plan.getSecurity() != null) {
+            throw new PlanInvalidException("Security type is forbidden for plan with 'Push' mode");
+        }
+    }
+
+    public void validatePlanTagsAgainstApiTags(Set<String> planTags, Set<String> apiTags) {
+        if (!isEmpty(planTags) && (isEmpty(apiTags) || apiTags.stream().noneMatch(planTags::contains))) {
+            log.debug("Plan rejected, tags {} mismatch the tags defined by the API ({})", planTags, apiTags);
+
+            throw new ValidationDomainException(
+                "Plan tags mismatch the tags defined by the API",
+                Map.of("planTags", String.join(",", planTags), "apiTags", String.join(",", apiTags))
+            );
+        }
+    }
+
+    public void validateGeneralConditionsPageStatus(Plan plan) {
+        if (plan.getGeneralConditions() != null && (plan.isPublished() || plan.isDeprecated())) {
+            var page = pageCrudService.findById(plan.getGeneralConditions());
+            var isPublished = page.map(Page::isPublished).orElse(false);
+            if (!isPublished) {
+                throw new ValidationDomainException("Plan references a non published page as general conditions");
+            }
+        }
+    }
+
+    private void ensurePlanSecurityIsAllowed(String securityType, String currentOrganizationId, String currentEnvironmentId) {
+        PlanSecurityType planSecurityType = PlanSecurityType.valueOfLabel(securityType);
+        Key securityKey =
+            switch (planSecurityType) {
+                case API_KEY -> Key.PLAN_SECURITY_APIKEY_ENABLED;
+                case KEY_LESS -> Key.PLAN_SECURITY_KEYLESS_ENABLED;
+                case JWT -> Key.PLAN_SECURITY_JWT_ENABLED;
+                case OAUTH2 -> Key.PLAN_SECURITY_OAUTH2_ENABLED;
+            };
+        if (
+            !parametersQueryService.findAsBoolean(
+                securityKey,
+                new ParameterContext(currentEnvironmentId, currentOrganizationId, ParameterReferenceType.ENVIRONMENT)
+            )
+        ) {
+            throw new UnauthorizedPlanSecurityTypeException(planSecurityType.name());
+        }
+    }
+
+    public static boolean isEmpty(Collection<?> collection) {
+        return collection == null || collection.isEmpty();
+    }
```

Comment:
I know we are in `core` and we don't want to use any library, but we should consider having a `utils` package in core for those kind of checks
imho it does not belong to this service.